src/backends/plonky2/primitives/merkletree_circuit.rs

```diff
@@ -0,0 +1,484 @@
+        }
+
+        if !existence && proof.other_leaf.is_some() {
+            // non-existence case ii) expected leaf does exist but it has a different key
+            pw.set_bool_target(self.case_ii_selector, true)?;
+            pw.set_target_arr(&self.other_key, &proof.other_leaf.unwrap().0 .0.to_vec())?;
+            pw.set_target_arr(&self.other_value, &proof.other_leaf.unwrap().1 .0.to_vec())?;
+        } else {
+            // existence & non-existence case i) expected leaf does not exist
+            pw.set_bool_target(self.case_ii_selector, false)?;
+            pw.set_target_arr(&self.other_key, &EMPTY_VALUE.0.to_vec())?;
+            pw.set_target_arr(&self.other_value, &EMPTY_VALUE.0.to_vec())?;
+        }
+
+        Ok(())
+    }
+
+    fn compute_root_from_leaf(
+        builder: &mut CircuitBuilder<F, D>,
+        path: &Vec<BoolTarget>,
+        leaf_hash: &HashOutTarget,
+        siblings: &Vec<HashOutTarget>,
+    ) -> Result<HashOutTarget> {
+        assert!(siblings.len() <= MAX_DEPTH);
+        // Convenience constants
+        let zero = builder.zero();
+        let one = builder.one();
+
+        // Generate/constrain sibling selectors
+        let sibling_selectors = siblings
+            .iter()
+            .rev()
+            .scan(zero, |cur_selector, sibling| {
+                let sibling_is_empty = sibling.elements.iter().fold(builder._true(), |acc, x| {
+                    let x_is_zero = builder.is_equal(*x, zero);
+                    builder.and(acc, x_is_zero)
+                });
+                // If there is a sibling, the selector is true, else retain the
+                // current selector
+                *cur_selector = builder.select(sibling_is_empty, *cur_selector, one);
+                Some(BoolTarget::new_unsafe(*cur_selector))
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+            .collect::<Vec<_>>();
+
+        let mut h = leaf_hash.clone();
+        for (i, (sibling, selector)) in std::iter::zip(siblings, &sibling_selectors)
+            .enumerate()
+            .rev()
+        {
+            // to compute the hash, we want to do the following 3 steps:
+            //     Let s := path[i], then
+            //     input_1 = sibling * s + h * (1-s) = select(s, sibling, h)
+            //     input_2 = sibling * (1-s) + h * s = select(s, h, sibling)
+            //     new_h = hash([input_1, input_2])
+            // TODO explore if to group multiple muls in a single gate
+            let input_1: Vec<Target> = (0..4)
+                .map(|j| builder.select(path[i], sibling.elements[j], h.elements[j]))
+                .collect();
+            let input_2: Vec<Target> = (0..4)
+                .map(|j| builder.select(path[i], h.elements[j], sibling.elements[j]))
+                .collect();
+            let new_h = builder.hash_n_to_hash_no_pad::<PoseidonHash>([input_1, input_2].concat());
```

Comment:
In the spec the intermediate node is calculated like this:
https://github.com/0xPARC/pod2/blob/abce0af675ba707b6f07b36b44a72c77036bf5a3/book/src/merkletree.md?plain=1#L43
This implementation is missing the `2` as third input.  The rust implementation has the same discrepancy.

The idea behind padding with `1` and `2` was to avoid collisions when calculating a leaf and intermediate node, so that an attacker can't reuse an intermediate node as a leaf or vice versa.  I think the current implementation also protects against this attack; but we should keep the implementation and spec in sync, so we need to choose one option.

Reply:
Right, in this case the circuit is implemented to mimic the rust impl. But seeing now the spec, I think it makes sense to go with the spec approach, so that hashing a leaf and an intermediate node takes the same amount of inputs. This is not specifically relevant for when using plonky2, but might be relevant for other proving systems so that the same hash circuit/chip is reused both for leaf & intermediate node. Will update the rust & circuit impls of the tree with that.